lib/Dialect/TTIR/Transforms/TTIRFusing.cpp

```diff
@@ -1055,6 +1055,130 @@ class PadPoolingFusionPattern : public mlir::OpRewritePattern<PoolingOp> {
   }
 };
 
+// Fuse MatmulOp followed by AddOp into a single LinearOp.
+//// This pattern looks for an AddOp where one of its operands is the result of
+// a MatmulOp and the other operand is a bias term. It then replaces the
+// AddOp with a LinearOp that combines the functionality of both operations.
+//// The pattern also handles the case where the MatmulOp is followed by a
+// ReshapeOp before the AddOp. In this case, it reshapes the bias term
+// accordingly and creates a LinearOp followed by a ReshapeOp to maintain the
+// original output shape.
+class MatmulWithBiasFusionPattern : public mlir::OpRewritePattern<AddOp> {
+public:
+  using mlir::OpRewritePattern<AddOp>::OpRewritePattern;
+  mlir::LogicalResult
+  matchAndRewrite(AddOp addOp, mlir::PatternRewriter &rewriter) const final {
+    if (MatmulOp matmulOp = isFusableMatmulOp(addOp); matmulOp) {
+      Value bias = addOp.getLhs() == matmulOp.getResult() ? addOp.getRhs()
+                                                          : addOp.getLhs();
+      Value matmulOpA = matmulOp.getA();
+      Value matmulOpB = matmulOp.getB();
+      LinearOp linearOp = rewriter.create<LinearOp>(
+          addOp.getLoc(), addOp.getResult().getType(), matmulOpA, matmulOpB,
+          bias, addOp.getOutput(), matmulOp.getTransposeA(),
+          matmulOp.getTransposeB());
+
+      rewriter.replaceOp(addOp, linearOp);
+
+      return mlir::success();
+    }
+    if (MatmulOp matmulOp = isFusableReshapeMatmulOp(addOp); matmulOp) {
+
+      ReshapeOp reshapeOp =
+          mlir::dyn_cast<ReshapeOp>(*matmulOp.getResult().getUsers().begin());
+      assert(reshapeOp &&
+             "MatmulWithBiasFusionPattern isFusable logic is broken.");
+
+      Value bias = addOp.getLhs() == reshapeOp.getResult() ? addOp.getRhs()
+                                                           : addOp.getLhs();
+
+      Value matmulOpA = matmulOp.getA();
+      Value matmulOpB = matmulOp.getB();
+      llvm::ArrayRef<int64_t> matmulOpShape =
+          matmulOp.getResult().getType().getShape();
+      llvm::ArrayRef<int64_t> addOpShape =
+          addOp.getResult().getType().getShape();
+      SmallVector<int32_t> matmulShapeI32(matmulOpShape.begin(),
+                                          matmulOpShape.end());
+      SmallVector<int32_t> addShapeI32(addOpShape.begin(), addOpShape.end());
+
+      auto biasType = mlir::cast<RankedTensorType>(bias.getType());
```

Comment:
TTIR dialect uses `RankedTensorType` so there shouldn't be a need of a typecast.
```suggestion
      RankedTensorType biasType = bias.getType();
```